### PR TITLE
Add promoter component

### PR DESCRIPTION
**Pull request by JoelSpeed**
Promoter is used as an external plugin to Prow which takes a source and target branch as query parameters and, when a PR is merged to the source branch, will automatically create a PR to promote the contents of the source branch to the production branch.

To enable this on a repo add an external plugin config:
```
- name: promoter
    events:
    - pull_request
    endpoint: http://promoter?source=<source>&target=<target>
```

### CODE REVIEW

File: plugins/promoter/server.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+	}
+
+	return nil
+}
+
+func (s *Server) createComment(org, repo string, num int, resp string) error {
+	return s.ghc.CreateComment(org, repo, num, fmt.Sprintf("Error occurred promoting branch: %s", resp))
+}
+
+func inSlice(list []string, toFind string) bool {
```
Comment (mthssdrbrg) [inline plugins/promoter/server.go]: `s/inSlice/contains`?

File: plugins/promoter/server.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+		return fmt.Errorf("cannot push without credentials - configure your git client")
+	}
+	s.log.Infof("Pushing to '%s/%s (branch: %s)'.", org, repo, branch)
+	remote := fmt.Sprintf("https://%s:%s@%s/%s/%s", s.botName, s.getBotPass(), "github.com", org, repo)
+	co := exec.Command(s.git, "push", remote, branch)
+	co.Dir = dir
+	out, err := co.CombinedOutput()
+	if err != nil {
+		s.log.Errorf("Pushing failed with error: %v and output: %q", err, string(out))
+		return fmt.Errorf("pushing failed, output: %q, error: %v", string(out), err)
```
Comment (mthssdrbrg) [inline plugins/promoter/server.go]: Any particular reason why these two errors have different messages?

File: plugins/promoter/server.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+	params := r.URL.Query()
+
+	sources, ok := params["source"]
+	if !ok {
+		responseHTTPError(w, http.StatusBadRequest, "400 Bad Request: Missing source parameter")
+		return []string{}, []string{}, false
+	}
+
+	targets, ok := params["target"]
+	if !ok {
```
Comment (mthssdrbrg) [inline plugins/promoter/server.go]: Shouldn't we be checking whether the `source` and `target` params are non-empty as well?

File: plugins/promoter/server.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+	return pluginHelp, nil
+}
+
+// Server implements http.Handler. It validates incoming GitHub webhooks and
+// then dispatches them to the appropriate plugins.
+type Server struct {
+	tokenGenerator   func() []byte
+	botName          string
+	botPassGenerator func() []byte
+	email            string
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply plugins/promoter/server.go]: It isn't actually used anywhere, so I've removed it 👍

File: plugins/promoter/server.go (hunk 3 above)
Comment (JoelSpeed) [reply plugins/promoter/server.go]: Added a check

File: plugins/promoter/server.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+	for _, target := range targets {
+		err = s.createPromotionPR(l, org, repo, target, body, author, title, num)
+		if err != nil {
+			errs = append(errs, err.Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		errors := strings.Join(errs, ", ")
+		return fmt.Errorf("error(s) encountered creating PR: %s", errors)
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply plugins/promoter/server.go]: Fixed

File: plugins/promoter/server.go (hunk 1 above)
Comment (JoelSpeed) [reply plugins/promoter/server.go]: Renamed

File: plugins/promoter/server.go (hunk 2 above)
Comment (JoelSpeed) [reply plugins/promoter/server.go]: They were copy pasted IIRC, so no, fixed

File: plugins/promoter/server.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mthssdrbrg) [reply plugins/promoter/server.go]: Even better! :sweat_smile:

**Pull request merged**